app/gui/controller/double-representation/src/name/project.rs

```diff
@@ -28,6 +28,88 @@ pub const STANDARD_BASE_LIBRARY_PATH: &str = concatcp!(STANDARD_NAMESPACE, ".",
 
 
 
+// ================
+// === Template ===
+// ================
+
+#[allow(missing_docs)]
+#[derive(Copy, Clone, Debug, Fail)]
+pub enum InvalidTemplateName {
+    #[fail(display = "The template name contains invalid characters.")]
+    ContainsInvalidCharacters,
+}
+
+/// The project template name.
+#[derive(Clone, Debug)]
+pub struct Template {
+    name: String,
+}
+
+impl Template {
+    /// Create the project template from string.
+    pub fn from_text(text: impl AsRef<str>) -> FallibleResult<Self> {
+        if text.as_ref().contains(|c: char| c.is_ascii_alphanumeric()) {
+            Ok(Template { name: text.as_ref().to_owned() })
+        } else {
+            Err(InvalidTemplateName::ContainsInvalidCharacters.into())
+        }
+    }
+
+    /// Create the project template from string without validation.
+    pub fn unsafe_from_text(text: impl AsRef<str>) -> Self {
+        Template { name: text.as_ref().to_owned() }
+    }
+
+    /// Create a project name from the template name.
+    pub fn to_project_name(&self) -> String {
+        let mut name = self.name.to_string();
+        // Capitalize
+        if let Some(r) = name.get_mut(0..1) {
+            r.make_ascii_uppercase();
+        }
+
+        name
+    }
+}
```

Comment:
Please add some tests for above code. Perhaps a simple doc tests will be enough.